rfcs/table-destructuring.md

```diff
@@ -0,0 +1,236 @@
+# Table Destructuring
+
+## Summary
+A feature present and highly-used in JavaScript, table destructuring allows the developer to extract data from tables and assign them to variables in a clean, shortened manner, thus avoiding the need for excessive index calls and several individual assignments, most of which being expendable.
+
+
+## Motivation
+As of today, the only way to "dynamically" assign table values is via `unpack`/`select` and variable list assignments. With that being said, these functions cannot be used on dictionaries and limit control over what values are being retrieved.
+
+With the introduction of destructuring assignments, we'll be able to substantially reduce redundant variables, index calls, and bloat in our code. Instead, using a clean syntax that allows the intuitive selection and assignment of values from arrays and dictionaries. 
+
+Destructuring allows for shortened code with increased flexibility. The expected outcome__s__ are endless.
+
+```lua
+-- Example using HTTP Service
+-- Based on example from https://developer.roblox.com/en-us/api-reference/class/HttpService
+
+local response = HttpSevice:GetAsync("http://api.open-notify.org/astros.json")
+local data = HttpService:JSONDecode(response)
+
+local { message, number, people } = data
+
+if message == "success" then
+    print("There are currently", data.number, "astronauts in space:")
+    for i, { name, craft } in people do
+        print(i .. ": " .. name .. " is on " .. craft)
+    end
+end
+```
+```lua
+-- Example using Fusion
+
+local { New, Children } = Fusion
+
+return New "ScreenGui" {
+    -- ...
+}
+```
+```lua
+-- Example using a RemoteFunction callback
+
+RemoteFunction.InvokeServer = function(player, { target, damage })
+    print("Dealt " .. damage .. " to " .. target.Name)
+end
+```
+
+## Design
+
+### Variable Assignments for Arrays
+With the ability to use `unpack` to destructure arrays, the need for this feature can seem negligible. Nevertheless, `unpack` cannot be used on dictionaries.
+```lua
+-- Current
+local foo = {1, 2, 3}
+
+local one, two, three = unpack(foo) -- 1, 2, 3
+```
+```lua
+-- New
+local foo = {1, 2, 3}
+
+local {one, two, three} = foo -- 1, 2, 3
+```
+
+### Variable Assignments for Dictionaries
+```lua
+-- Current
+local foo = {
+    red = "good";
+    green = "better";
+    blue = "best";
+}
+
+local red, green, blue = foo.red, foo.green, foo.blue
+print(red) -- good
+print(green) -- better
+print(blue) -- best
+```
+```lua
+-- New
+local foo = {
+    red = "good";
+    green = "better";
+    blue = "best";
+}
+
+local {red, green, blue} = foo
```

Comment:
This section is semantically ambiguous. You can't use the same syntax and have the language somehow detect the author's intention.